SteamKit2/SteamKit2/Steam/Handlers/SteamMatchmaking/SteamMatchmaking.cs

```diff
@@ -0,0 +1,392 @@
+                    var members = existingLobby?.members;
+
+                    return new Lobby(
+                        lobbyId,
+                        ( ELobbyType )lobby.lobby_type,
+                        lobby.lobby_flags,
+                        existingLobby?.OwnerSteamID,
+                        Lobby.DecodeMetadata( lobby.metadata ),
+                        lobby.max_members,
+                        lobby.num_members,
+                        members?.Count == lobby.num_members - 1 ? members : null,
+                        lobby.distance,
+                        lobby.weight
+                    );
+                } );
+
+            foreach ( var lobby in lobbyList )
+            {
+                appLobbies[ lobby.SteamID ] = lobby;
+            }
+
+            Client.PostCallback( new LobbyListCallback(
+                body.app_id,
+                ( EResult )body.eresult,
+                lobbyList
+            ) );
+        }
+
+        void HandleLobbyData( IPacketMsg packetMsg )
+        {
+            var lobbyListResponse = new ClientMsgProtobuf<CMsgClientMMSLobbyData>( packetMsg );
+            CMsgClientMMSLobbyData body = lobbyListResponse.Body;
+
+            var lobbyId = new SteamID( body.steam_id_lobby );
+            var appLobbies = lobbies.GetOrAdd( body.app_id, k => new ConcurrentDictionary<SteamID, Lobby>() );
+            var existingLobby = appLobbies.ContainsKey( lobbyId ) ? appLobbies[ lobbyId ] : null;
+
+            List<Lobby.Member> memberList =
+                body.members.ConvertAll( member => new Lobby.Member(
+                    new SteamID( member.steam_id ),
+                    member.persona_name,
+                    Lobby.DecodeMetadata( member.metadata )
+                ) );
+
+            var updatedLobby = new Lobby(
+                lobbyId,
+                ( ELobbyType )body.lobby_type,
+                body.lobby_flags,
+                new SteamID( body.steam_id_owner ),
+                Lobby.DecodeMetadata( body.metadata ),
+                body.max_members,
+                body.num_members,
+                memberList,
+                existingLobby?.Distance,
+                existingLobby?.Weight
+            );
+
+            appLobbies[ lobbyId ] = updatedLobby;
+
+            Client.PostCallback( new LobbyDataCallback( body.app_id, updatedLobby ) );
+        }
+
+        void HandleUserJoinedLobby( IPacketMsg packetMsg )
+        {
+            var userJoinedLobby = new ClientMsgProtobuf<CMsgClientMMSUserJoinedLobby>( packetMsg );
+            CMsgClientMMSUserJoinedLobby body = userJoinedLobby.Body;
+
+            var lobbyId = new SteamID( body.steam_id_lobby );
+            var userId = new SteamID( body.steam_id_user );
+
+            var appLobbies = lobbies.GetOrAdd( body.app_id, k => new ConcurrentDictionary<SteamID, Lobby>() );
+            var lobby = appLobbies.ContainsKey( lobbyId ) ? appLobbies[ lobbyId ] : null;
```

Comment:
This does two dictionary lookups. Better to use `TryGetValue` for only one lookup.